src/forms/compoundField.js

```diff
@@ -0,0 +1,153 @@
+const Joi = require('joi');
+const option = require('option');
+const FieldError = require('./fieldError');
+
+class BadValidationTargetError extends Error {
+  constructor(id, field) {
+    const targets = [field.name, ...field.fields.map(f => f.name)];
+    const message = `${id} not recognised. Must be one of ${targets.join(',')}`;
+    super(message);
+    Error.captureStackTrace(this, this.constructor);
+    this.name = this.constructor.name;
+    this.message = message;
+    this.targetId = id;
+  }
+}
+
+const isNullOrUndefined = value =>
+  typeof value === 'undefined' || value === null;
+
+const failOnFirstFailure = validations => {
+  if (!(validations && validations.length)) {
+    return { result: true, errors: [] };
+  }
+  const [currentValidation, ...rest] = validations;
+  const maybeError = currentValidation();
+
+  if (!maybeError || isNullOrUndefined(maybeError)) {
+    return failOnFirstFailure(rest);
+  }
+  return { result: false, errors: [maybeError] };
+};
+
+const errorFor = (id, message) => {
+  return { id, message };
+};
+
+const parseErrorTarget = (targetOrMessage, fallbackId) => {
+  const isObject = typeof targetOrMessage === 'object';
+  if (isObject && 'message' in targetOrMessage && 'id' in targetOrMessage) {
```

Comment:
would read slightly nicer and consistent if it was `isObject && hasMessage && hasId`

Reply:
Can't run the has message or hasId ones until I know it is an object though. Non objects will throw on an `in` check. Could make them functions.

Reply:
ok, I've added new checks `isObject`, `hasKey` and `kasKeys` to [src/utils/checks.js](https://github.com/hmcts/one-per-page/blob/compound-field/src/util/checks.js#L10-L17). That cleans up this line to:
```
if (isObject(targetOrMessage) && hasKeys(targetOrMessage, 'message', 'id')) {
```